Restore SMP to a working state, enable pervasively on x86_64

andyross opened the pull request:
SMP got kinda rotten across the timer and scheduler rework, as our only SMP platform was a hardware-only device (esp32) with somewhat quirky behavior in hardware test integration and no serious users within the community.

This series fixes a bunch of bugs that crept in, and a few that were never discovered originally (the kernel/smp test itself has been significantly enhanced since).  And it caps it off by making SMP the default on x86_64 (everywhere it works, anyway -- lots of tests aren't smp-safe by design).

Full disclosure: the interprocessor interrupt framework in the "Support synchronous k_thread_abort" patch probably constitutes a new feature from the perspective of the 1.14 code freeze.  But (1) this was much easier than trying to find all the uses of k_thread_abort() throughout the tests and "fix" them to work in SMP (where k_thread_abort() basically didn't work by design) and (2) if you look carefully, I believe it's true that there are ZERO changes in code paths when CONFIG_SMP=n.  It only affects the situation I'm fixing here.

Commit:
arch/x86_64: Clean up stack initialization

When this code was written, there was no "stack frame" struct defined.
There is now, so use that for clarity and concision.  Also remove an
obvious comment (I mean, duh, we can put any segment selectors in
those fields we want).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
drivers/timer/hpet: Add noop SMP initialization function

This is intended to initialize CPU-local timer devices, but HPET is
global so we have nothing to do.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
cmake/emu: Add qemu SMP flags automatically

No need to do this in arch code, all qemu variants speak the same
command line and the kconfig values we need are easily available.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/x86_64: board/x86_64/qemu_x86_64: Enable 2-processor SMP as default

We can turn this off in application code, but the default should be
SMP.

Oddball note: the SMP setting is in the soc layer and not arch, even
though I think it would be better there.  We don't actually have a
per-arch defconfig file included by kconfig.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
arch/x86_64: Add a per-CPU SMP bringup API

The previous scheme where the xuk layer would call out to the to
"fetch" the stack for a SMP CPU at startup was sorta weird, and an
impedance mismatch with Zephyr which has a "start this CPU" call
instead.  It also got broken when x86_64 started launching CPUs
(correctly) on their interrupt stacks instead of a temporary area;
they weren't ready yet when xuk initialization was happening and the
system would deadlock waiting for code that can't run yet to provide a
stack.

Note that this preserves the somewhat quirky behavior that Zephyr's
CPU numbering is just the order in which the SMP CPUs emerge from
initialization and not a hardware ID.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/smp: Warning cleanup

When CONFIG_SMP is enabled but CONFIG_MP_NUM_CPUS is 1 (which is a
legal configuration, though a weird one) this static function ends up
being defined but unused, producing a compiler warning.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Handle aboring _current correctly in SMP

In SMP, _current is not "queued".  (The run queue only stores
unscheduled threads because we can't rely on the head of the list
being _current).  We weren't updating the cache choice, which would
flag swap_ok, so calling k_thread_abort(_current) (for example, when a
thread exits from its entry function) would try to switch back into
the thread and then run off the end of the function.

Amusingly this was more benign than you'd think.  Stumbled on it by
accident.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
x86_64: Missing a volatile on a struct used for spinning

Before we're initialized and can use proper synchronization, the CPU
initialization path spins on the thread entry function to be non-null.
But the data wasn't tagged volatile, and with gcc 8.2.1 (but not
6.2.0) the optimizer was hoisting the reads to SMP init would spin
forever.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/smp: Fix bitrot with the way the SMP "start flag" works

Tickless timers mean that k_busy_wait() won't work until after the
timer driver is initialized, which is very early but not as early as
SMP.  No need for it, just spin.

Also the original code used a stack variable for the start flag, which
racily presumed that _arch_start_cpu() would comes back synchronously
with the other CPU fired up and running right now.  The cleaned up smp
bringup API on x86_64 isn't so perky, so it exposed the bug.  The flag
just needs to be static.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: arch/x86_64: Support synchronous k_thread_abort() in SMP

Currently thread abort doesn't work if a thread is currently scheduled
on a different CPU, because we have no way of delivering an interrupt
to the other CPU to force the issue.  This patch adds a simple
framework for an architecture to provide such an IPI, implements it
for x86_64, and uses it to implement a spin loop in abort for the case
where a thread is currently scheduled elsewhere.

On SMP architectures (xtensa) where no such IPI is implemented, we
fall back to waiting on an arbitrary interrupt to occur.  This "works"
for typical code (and all current tests), but of course it cannot be
guaranteed on such an architecture that k_thread_abort() will return
in finite time (e.g. the other thread on the other CPU might have
taken a spinlock and entered an infinite loop, so it will never
receive an interrupt to terminate itself)!

On non-SMP architectures this patch changes no code paths at all.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Clean up tracing hooks

The tracing fixes in commit e87193896a83 ("subsys: debug: tracing: Fix
thread tracing") were... not a readability win.  The point appears to
have been to put a tracing hook immediately before and after the
assignment to the _current pointer.  So do that in an abstracted
function and clean up _get_next_switch_handle() (which is a subtle and
important function already polluted with some unavoidable preprocessor
testing!)

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/smp: Honor TEST_EXTRA_STACKSIZE

There was a test-created thread that wasn't including this.  It's a
huge stack and doesn't overflow (though I thought briefly that it
was), but it's a rule that we need to have that buffer and I'm trying
to fix these as I find them.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Fix spinlock validation glitch in SMP

In SMP, we are setting the _current pointer while holding the
scheduler spinlock locally, which means that when we try to release it
the validation layer (not the spinlock per se) will scream at us
because the thread that took the lock doesn't match the one releasing
it.

Special case this when validation is enabled.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/spinlock: Fix race in spinlock validation

The k_spin_lock() validation was setting the new owner of the spinlock
BEFORE the actual lock was taken, so it could race against other
processors trying the same thing.  Split the modification step out
into a separate function that can be called after we affirmatively
have the lock.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Fix k_yield() in SMP

This was always doing a remove/add of the _current thread to the run
queue, which is wrong because in SMP _current isn't in the queue to
remove.  But it went undetected until the recent dlist changes.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/smp: Remove test_wakeup_pending_threads case

This case was predicated on a mistake.  The behavior of k_wakeup() has
always been NOT to wake up threads that are "pending" on a wait queue,
only ones blocked on a timeout in k_sleep().  As written, this test
case could never pass.

(Really there's no good reason for that.  It seems reasonable to me to
expect wakeup to work symmetrically, and the docs are sort of
ambiguous on the subject.  But the code in k_wakeup() is clear:
threads flagged pending get an early exit and the call becomes a
noop.)

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/smp: Clean up "guess waiting" on SMP thread exit

The various tests would all do a "wait for threads to exit" step
before checking the results, but this was implemented with a simple
busy wait that turns out to need careful tuning (because there was
busy waiting in the threads).

Rather than try to synchronize this, white box the issue (it's a low
level SMP test, after all) by spinning on the thread states directly
watching for the kernel to flag them dead.  The downside here is that
if the process fails for some reason we'll get a hang and a timeout
reported from sanitycheck and not a synchronous ztest assertion.  But
in return, successful tests run much faster and I don't need to worry
about how to tune them for IPI latency on different platforms.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

dcpleung commented on drivers/timer/hpet.c:
```diff
@@ -98,6 +98,14 @@ int z_clock_driver_init(struct device *device)
 	return 0;
 }
 
+void smp_timer_init(void)
```
Might want to wrap this inside "#ifdef CONFIG_SMP"

andyross replied:
Not sure I see why; if it's not referenced it won't get pulled into the link.

dcpleung replied:
I was thinking "defined but no used" warning. Come to think of it, it shouldn't matter.

dcpleung commented on kernel/smp.c:
```diff
@@ -12,6 +12,7 @@
 
 #ifdef CONFIG_SMP
 static atomic_t global_lock;
+static atomic_t start_flag;
```
I pulled in this patch for my own branch and it failed some sanitychecks. This is defined within CONFIG_SMP, but the usage inside smp_init() below is not inside a #ifdef. So there were errors.

andyross replied:
Seems like it is to me and it builds on both SMP=y/n in CI.  Can you post the error?

dcpleung replied:
> zephyr/kernel/smp.c: In function ‘smp_init’:
> zephyr/kernel/smp.c:96: error: ‘start_flag’ undeclared (first use in this function)
> zephyr/kernel/smp.c:96: error: (Each undeclared identifier is reported only once
> zephyr/kernel/smp.c:96: error: for each function it appears in.)
> make[2]: *** [zephyr/kernel/CMakeFiles/kernel.dir/build.make:310: zephyr/kernel/CMakeFiles/kernel.dir/smp.c.obj] Error 2

I was just building `samples/hello_world` with no modification. But this was under XCC.

Commit:
arch/x86_64: Flag xuk shared page volatile

The shared page is inherently used in multiprocessor contexts where
the compiler optimizer can trip us up (specifically, a spin on
num_active_pus was being hoisted out of the loop on some gcc's).  Put
the volatile declartion into the struct pointer itself instead of
relying on the code to get it right.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andrewboie commented:
Hmm
Do we have bugs open on the tests that explode when SMP is enabled?

andyross commented:
Threw a DNM tag on this because the PR now how a temporary patch in it to force the use of a particular SDK.   That needs to come out after CI is upgraded and before merge.

And indeed, I should open a bug to audit all those tests.  Most of them aren't really "bugs" -- just tests that didn't expect there to be another CPU.   I mean, the tests/kernel/sched test that I wrote is 100% predicated on the idea of throwing a bunch of threads at a single CPU scheduler and seeing what runs first.

Lots of that kind of thing can be fixed by changing the test code a little to e.g. put a high priority thread on the other CPU for the duration of the specific failing test cases, or by using the SMP affinity API to pin threads, etc...  And I'm sure at least a few of those failures reflect real bug with the SMP or arch layer too.

andyross commented:
Rebased & fixed the gitlint errors (that were on the DNM patch which can't be merged anyway, but might as well make it show green).  I'm a little at a loss for the build errors where we're overflowing code space on a handful of 32kb platforms.  Those same tests build with the same SDK for me locally.  They're tight, but they fit.  Is something extra turned on for CI builds?

andyross commented:
OK, figured out the timeouts.  The "modified tests" script is noticing that I've, heh, modified a ton of tests, so it's adding those directories (no shock, they'd get built anyway) **and** a "--all" to the sanitycheck argument, presumably because the tests need to be build-validated on all platforms?  And apparently CI can't currently handle an --all build in the two hours a pull request is given to finish?  Or plausibly the new SDK is slower than the old one by enough to push it over the edge.

@nashif -- what's the best way forward here?

andyross commented:
Note that the giant "disable SMP in a bunch of tests" patch has been split out into three patches that can be merged before this one (as without this PR, they're noops).  That should fix the CI build time troubles.

andrewboie commented:
needs rebase.

andyross commented:
Oops, pushed the version with the included squash patches.  One sec...
